Week1 [STEP 3] Dino

Jnnsu opened the pull request:
@1Consumption 

각각의 매개변수를 받는 함수를 만드는 과정까지는 어렵지 않았으나 매개변수를 이미 받은 함수를 다른 함수 내에 서 호출할 때, 매개변수를 어떻게 설정해 줘야하는 고민이 있었습니다. 
예를 들면
```Swift
func chooseBodyAndToppingOfIceCream(body: String, topping: String) {
    print(" \(topping)\(body)\(topping)")
}
```
이 함수를 만들어 놓고
```Swift
func makeIceCreamPart(lengthOfIceCream: Int) {
    for rowOfIceCream in 1...lengthOfIceCream {
        chooseBodyAndToppingOfIceCream(body: String, topping: String)
    }
}
```
이렇게 함수 내에 함수를 호출했을 때가 난감했습니다. 이렇게 함수들의 매개변수를 독립적으로 쓰게 되니, 마지막에 <정보>를 출력하는 과정에서도 컴파일 에러가 나타났습니다.
```Swift
 print("<정보>\n 길이: \(lengthOfIceCream)\n 몸통: \(body)\n 토핑: \(topping)\n 막대길이: \(lengthOfStick)\n")
// 에러났습니다..
```
고민해본 결과, 매개변수를 매개변수 값으로 호출해서 더 큰 함수에 매개변수를 이어서 받게 해보자는 생각으로 시도해봤습니다.
```Swift
func makeIceCreamPart(lengthOfIceCream: Int, body: String, topping: String) {
    for rowOfIceCream in 1...lengthOfIceCream {
        chooseBodyAndToppingOfIceCream(body: body, topping: topping)
    }
}
```
이렇게 바꾸었더니 컴파일 에러가 나지 않고 잘 작동하는 것을 확인했습니다!
이런 방법으로 <정보>를 출력하는 메소드도 함수 내에 묶어버리니 해결됐습니다.
```Swift
func makeWholeIceCream(lengthOfIceCream: Int, body: String, topping: String, lengthOfStick: Int) {
    print("<정보>\n 길이: \(lengthOfIceCream)\n 몸통: \(body)\n 토핑: \(topping)\n 막대길이: \(lengthOfStick)\n")
    makeIceCreamPart(lengthOfIceCream: lengthOfIceCream, body: body, topping: topping)
    makeStickPart(lengthOfStick: lengthOfStick)
}
```

Jnnsu committed:
Week1_Step2_Dino.swift 파일 추가

Jnnsu committed:
main.swift -> Week1_Step2.swift 파일명 변경

Jnnsu committed:
Week1 [STEP 3] Dino

Jnnsu committed:
Week1 [STEP 3] Dino count -> rowOfIceCream, rowOfStick 변경

Jnnsu committed:
Week1 [STEP 3] Dino flowChart 파일 추가

Jnnsu committed:
Week1 [STEP 3] Dino xcodeproj 커밋

1Consumption commented on CodeStarterCamp_Week1/Week1_Step2.swift:
```diff
@@ -0,0 +1,47 @@
+}
+makeIceCream()
+makeStick()
+*/
```
주석으로 코드를 남기기보다는 이미 github에 올라가있으니 지우는게 어떨까요? 추후에 참고가 필요하다면 이전 커밋을 참고하면 되니까요.

Jnnsu replied:
넵 불필요한 부분은 삭제하겠습니다!

1Consumption commented on CodeStarterCamp_Week1/Week1_Step2.swift:
```diff
@@ -0,0 +1,47 @@
+//  CodeStarterCamp_Week1
+//
+//  Created by yagom.
+//  Copyright © yagom academy. All rights reserved.
```
이 파일의 이름이 main.swift인가요?

Jnnsu replied:
이름이 잘못되었군요.. 수정하겠습니다.

1Consumption commented on CodeStarterCamp_Week1/main.swift:
```diff
@@ -1,47 +1,39 @@
-    }
+func chooseBodyAndToppingOfIceCream(body: String, topping: String) {
+    print(" \(topping)\(body)\(topping)")
 }
```
메소드 이름에 and가 들어가는건 좋지 못합니다. 여러개의 역할을 하는 이름 같네요. 조금 더 추상화를 해보는건 어떨까요?

Jnnsu replied:
Body와 Topping을 모두 포함할 수 있는 이름이 좋겠군요. 더 이해하기 쉽고 한가지 역할을 하는 이름으로 바꿔봐야겠네요

1Consumption commented on CodeStarterCamp_Week1/main.swift:
```diff
@@ -1,47 +1,39 @@
+    for rowOfIceCream in 1...lengthOfIceCream {
+        chooseBodyAndToppingOfIceCream(body: body, topping: topping)
     }
 }
```
0이 아닌 1부터 시작하는 이유가 있을까요?

Jnnsu replied:
0이 아닌 1부터 시작하게 한 이유는 입력하는 매개변수 값과 출력되는 값이 같게하기 위해서였습니다. 
0부터 시작했을 경우 lengthOf~매개변수에 n값을 넣으면 n+1 만큼 반복할 것이라고 생각했습니다.

1Consumption commented on CodeStarterCamp_Week1/main.swift:
```diff
@@ -1,47 +1,39 @@
+func makeWholeIceCream(lengthOfIceCream: Int, body: String, topping: String, lengthOfStick: Int) {
+    print("<정보>\n 길이: \(lengthOfIceCream)\n 몸통: \(body)\n 토핑: \(topping)\n 막대길이: \(lengthOfStick)\n")
+    makeIceCreamPart(lengthOfIceCream: lengthOfIceCream, body: body, topping: topping)
+    makeStickPart(lengthOfStick: lengthOfStick)
```
이 두 메소드에 들어가는 lengthOf~ 매개변수는 이미 메소드의 이름에서 어떤걸 만드는지 표현이 되었으니 length 정도로 표시하는건 어떨까요?

Jnnsu replied:
더 간단한 표현방법이 있었네요..!! 이 부분도 따로 수정하겠습니다.

1Consumption commented on CodeStarterCamp_Week1/main.swift:
```diff
@@ -1,47 +1,39 @@
 }
-makeStick()
+
+//실행 예시 1
+makeWholeIceCream(lengthOfIceCream: 10, body: " ***", topping: "", lengthOfStick: 4)
+//실행 예시 2
+makeWholeIceCream(lengthOfIceCream: 12, body: "***", topping: "&", lengthOfStick: 4)
+//실행 예시 3
+makeWholeIceCream(lengthOfIceCream: 12, body: "***", topping: "#", lengthOfStick: 6)
+//실행 예시 4
+makeWholeIceCream(lengthOfIceCream: 6, body: " |0|", topping: "", lengthOfStick: 4)
```
매개변수만 바꾸면 여러가지 빼빼로가 나오는 멋진 메소드군요~

Jnnsu replied:
좋게 봐주셔서 감사합니다~😄

1Consumption reviewed (changes requested):
안녕하세요 디노님~ 리뷰가 늦어져서 죄송합니다 ㅠㅠ 😢 

올려주신 PR 잘 봤어요! 발전한게 느껴집니다 👍👍👍
커밋을 나누려고 노력한 흔적이 보이고, 잘 나누셨습니다.
네이밍도 신경쓴 부분이 많이 보였어요!

몇가지 아쉬운 부분도 있었는데,  아래에 적어볼게요.
1. 이 커밋은 이름만 보고는 어떤 내용인지 전혀 모르겠군요😢 커밋을 하기전에는 항상 확인하는 습관을 들여야합니다.
<img width="248" alt="image" src="https://user-images.githubusercontent.com/37682858/154207269-7f7eee63-077c-4720-a451-e4d2e68c48ef.png">



2. [Week1 [STEP 3] Dino xcodeproj 커밋](https://github.com/yagom-academy/swift-starter-week1/pull/211/commits/165f6f6073b9e7119e7585caeb7460691588811b) xcode는 project.pbxproj 의 구조를 바탕으로 파일을 트래킹 하는데, 이 파일에 대한 변경사항을 마지막에 한번에 몰아서 하면 중간 커밋에서 컴파일이 안되는 경우가 발생합니다. 고로 파일을 생성 또는 이름을 변경했다면 해당 변경에 해당하는 project.pbxproj 파일도 같은 커밋에 묶어줘야합니다.

3. rowOfIceCream의 값은 어떻게 되나요? 0부터 시작인가요 1부터 시작인가요? 이 부분도 플로우 차트에 포함되면 좋겠습니다. 그리고 만약 1부터 시작한다면 `rowOfIceCream이 lengthOfIceCream 보다 작은가`, `rowOfStick이 lengthOfStick보다 작은가` 이 두 조건은 `작거나 같은가`로 수정되어야 할 거 같습니다. 
<img width="430" alt="image" src="https://user-images.githubusercontent.com/37682858/154108076-08eef6cc-80ec-4776-8cc2-bc816a5ab722.png">

나머지는 코멘트를 참고해주세요!

Jnnsu committed:
Week1 [STEP 2] Dino 불필요한 주석 삭제

Jnnsu committed:
Week1 [STEP 2] Dino //main.swift -> //Week1_Step2.swift

Jnnsu committed:
Week1 [STEP 3] Dino main.swift의 //Week1_Step2_Dino -> //main.swift

Jnnsu committed:
Week1 [STEP 3] Dino chooseBodyAndToppingOfIceCream -> chooseOptionsOfIceCream

Jnnsu committed:
Week1 [STEP 3] Dino
makeIceCreamPart(lengthOfIceCream:)->makeIceCream(length:)
makeStickPart(lengthOfStick:)->makeStickPart(length:)

Jnnsu committed:
Week1_Step3_flowChart 삭제

Jnnsu committed:
flowChart2 -> Week1_Step2_flowChart 파일명 변경

Jnnsu committed:
Week1_Step3_flowChart 파일 추가
기존 flowChart에서 rowOfIceCream: 1, rowOfStick: 1 정보 추가됨

Jnnsu committed:
기존 Week1_Step3_flowChart.png 삭제

Jnnsu committed:
Week1_Step3_flowChart 피드백 후 수정: rowOf~값과 length값을 비교하기 전에 출력 -> 후에 출력
